docs/assignment3.md

```diff
@@ -1 +1,577 @@
+| **Intended use**       | When a player sends a command request to the server, it will go through the chain-of-responsibility. Depending on run-time conditions, the correct handler will be chosen. Let's take for example the `shout` command, the arguments of the command should be sent by the server to all other receivers as a broadcasted message. The Player class will receive the command, as it is the head, and forward it to the PlayerThread. It is not their responsibility to handle it, so it will keep forwarding it to the server. The server can then call the respective method to handle this request. <br> Another example is when the player joins the realm for the first time. It is then the PlayerThread's responsibility to receive the request and handle the receiving of the player's username. <br>When an incorrect command is sent over the chain, the request will end up in the last part, in our case the server. The server has a handler for incorrect requests and will reply with an error message. |
+| **Constraints**        | N/A |
+
+| ID                     | DP4 |
+| ---------------------- | ------------------------------------------------------------ |
+| **Design pattern**     | Decorator Pattern |
+| **Problem**            | The current game as-is provides creation of areas, items and obstacles via the input JSON file. Using the built-in commands a player can traverse over the areas, pickup or drop items, use the items on other objects and more. All functionality for the original game design is implemented, the storyline can be altered via the JSON. However, it can be understandable that a different developer or modder wants to add additional functionality – read: commands – to the game, to extend beyond possible in the JSON. Not everyone has the same level of experience with coding, which could be a problem. We should provide a flexible solution to add commands to the game without altering the source code. |
+| **Solution**           | A solution to this would be implementing an extended (decorated) instruction parser. This object should forward any current instructions to the default instruction parser in the game, but, when a different instruction arrives, it should parse it on its own. This parsing object could be initialized via the JSON file, making it easy to create functions for non-programmers. |
+| **Intended use**       | A use-case could be, for example, the addition of a `destroy` command, with syntax `destroy <item>`. The item would be removed from the backpack and, instead of being dropped, be completely gone. For the current game story that might not be desirable, but that's open and up to the developer. |
+| **Constraints**        | This pattern is not yet implemented but could serve as an idea for future versions. The commands are also based on only currently implemented commands. There are a lot of combinations to be made, but to get deeper functionality, the actual code needs to be edited. |
+
+## Class diagram									
```

Comment:
Nice design choice of keeping the multiplayer setting in mind when developing the diagrams